### ISSUE TITLE
[idea] fnv1a hash and replication (not consistent hash)

### ISSUE THREAD

**Issue by zzl0** [grobian/carbon-c-relay#108]
Hi

Thanks for your work.

I want add a feature that do fnv1a hash and replication, Following is an example configuration:

```
cluster HashAndReplicate
    fnv1a_h_r
        [cache-instance-0a cache-instance-0b]
        [cache-instance-1a cache-instance-1b]
        [cache-instance-2a cache-instance-2b]
        ...
        [cache-instance-15a cache-instance-15b]

    ;
# cache-instance-0a can be something like host:port=0a
```

When the cluster receive a metric
1. Compute the hash value, `h = fnv1a(metric_name)`
2. Compute the index of cache group, `idx = h % 16`
3. Send metric to all the cache instances in the group.

Can you help me review this idea, is there any better way to do this? 
I have tried [another way](https://github.com/zzl0/carbon-c-relay/commit/a224839c8cd9bb3554255022ddea282bf4a8cab5) to do fnv1a hash and replication, but I feel it's ugly to use `cluster` as  group.

Many thanks!

**Comment by grobian** [comment]
It feels like you want to create an exact copy of all stores, why is this so important for you?
I think if you want it this way, the better way would be to allow using duplicate keys, such that it becomes implicit, like so:
```cluster foo fnv1a_ch ip1=hash1 ip2=hash1 ip3=hash2 ip4=hash2;```
internally we'd have to find a way to make the mapping respected on the routing layer

**Comment by zzl0** [comment]
> It feels like you want to create an exact copy of all stores, why is this so important for you?

I want to know the metric position easily, and control number of replicas.

Thanks for your reply.

The reason why I want to know the metric position easily is that we had replace `whisper` with `kenshin`, a storage component to reduce  IO operations, it combined multiple metrics to a file.

> I think if you want it this way, the better way would be to allow using duplicate keys, such that it becomes implicit, like so:
> cluster foo fnv1a_ch ip1=hash1 ip2=hash1 ip3=hash2 ip4=hash2;

Do you mean `cluster foo fnv1a_ch replication 2 ip1=hash1 ip2=hash1 ip3=hash2 ip4=hash2;` ?

**Comment by grobian** [comment]
Interesting!  Do you have a link or something for kenshin? I'm planning to write something which sounds very similar.

if you'd add replication 2 to the cluster definition, all metrics would go to all four nodes

**Comment by zzl0** [comment]
> Interesting! Do you have a link or something for kenshin? I'm planning to write something which sounds very similar.

I'm glad to hear you are planning to do similar thing. [This](https://github.com/zzl0/Kenshin) is a personal project link, we plan to open source it, but now I need to write maintain tools and refactor code.

Initial ideas:
1. ~~Add `tag` to metric.~~
2. Merge multiple metrics.
3. Make xff (xFilesFactor) > 1

My implementation is based on `whisper` and `carbon-cache`.

looking forward to your reply :)

**Comment by grobian** [comment]
I was thinking about this, and it occurred to me that unless you want replications on individual stores (which you don't seem to do) you could just do it with two clusters, e.g.:
```
cluster a-variant fnv1a_ch
    host1=key1
    host2=key2
    host3=key3
    ;
cluster b-variant fnv1a_ch
    host4=key1
    host5=key2
    host6=key3
    ;
match *
    send to
        a-variant
        b-variant
    ;
```

**Comment by grobian** [comment]
I got the impression you guys fixed this for your use-case in your own branch, and in a generic way I think it can be expressed in the current relay implementation.  So I'm going to close this issue for I'm not sure there is anything to do here.

**Comment by zzl0** [comment]
@grobian yeah, thanks for your help, and sorry for the late reply.
Here is why we implemented our own branch:
1. We have used a in-house fnv1a implementation.
2. We want the metrics reported by relay are not cumulative. ([this pr](https://github.com/grobian/carbon-c-relay/pull/89) is rejected)
3. I don't want consistent hash now (for Kenshin).
4. Our current implementation compute the hash value only once. Thanks for your advice again, my first implementation is ugly
